### PR TITLE
library: Move `CStr` to libcore, and `CString` to liballoc

### LINKED ISSUE #46736: Allow CStr and CString types to be used with no_std
Embedded code may interact with C and thus may find these types useful. It'd make sense to allow usage of `CStr` in libcore and `CString` in liballoc.

Of course, right now due to the way implementations work in the standard library it may not be possible to include `CStr` in libcore. However, moving it to liballoc is still a step up.

### PR DESCRIPTION
**Pull request by petrochenkov**
Closes https://github.com/rust-lang/rust/issues/46736

Interesting points:
- Stability:
    - To make `CStr(ing)` from libcore/liballoc unusable without enabling features I had to make these structures unstable, and reexport them from libstd using stable type aliases instead of `pub use` reexports. (Because stability of `use` items is not checked.)
- Relying on target ABI in libcore is ok:
    - https://github.com/rust-lang/rust/pull/94079#issuecomment-1044263371
- `trait CStrExt` (UPDATE: used only in `cfg(bootstrap)` mode, otherwise lang items are used instead)
    - https://github.com/rust-lang/rust/pull/94079#issuecomment-1047863450
- `strlen`
    - https://github.com/rust-lang/rust/pull/94079#issuecomment-1047863450

Otherwise it's just a code move + some minor hackery usual for liballoc in `cfg(test)` mode.

### CODE REVIEW

**Comment by joshtriplett** [comment]
This is great! I'd love to see this move, to make it easier to interface with C FFI in contexts that don't necessarily want std.

Platform `strlen` implementations typically have extensive optimizations, and I don't think we'd want to miss out on those. It might make sense to ensure we have a weak strlen in compiler-builtins or similar. (cc @Amanieu)

**Comment by petrochenkov** [comment]
### Relying on target ABI in libcore is ok

A lot of attention in https://github.com/rust-lang/rust/issues/46736 is put on whether depending on `c_char` in libcore is an acceptable thing or not.
I don't think the self-imposed restriction of not doing that is reasonable.

Obviously, libcore cannot depend on calling OS services, be it syscalls or functions or system libraries.
But why cannot it rely on the target's ABI (which includes sizes/alignments/signedness of basic C types)? It doesn't introduce any new dependencies to libcore.

Even without libcore the compiler cares about the target's ABI and keeps the relevant data in target specifications (inside rustc and LLVM).
It must know these things to generate layouts for `repr(C)` data structures, and calling conventions for generating `extern "C"` functions.
rustc's target specs, for example, have fields like `c_int_width` and `c_enum_min_bits`, even if we don't count the LLVM part.
So libcores build for different targets are already different, because the compiler uses that ABI knowledge when generating code.
If the compiler can do it, then why a higher layer like libcore cannot?

Recently core contains more of target-dependent stuff than back in 2017 when https://github.com/rust-lang/rust/issues/46736 was created, because it's simply too useful to break that restriction sometimes.
A bunch of code in `core::arch` depends on the target's CPU architecture, and the `va_list` implementation in `core::ffi` requires relying on even more ABI details.

I personally don't see issues with putting that `cfg_if` that defines `c_char` into libcore, but perhaps people will feel better about this if the char signedness is moved to the compiler's target specs (`compiler\rustc_target\src\spec\*.rs`)?
Then the `c_char` definition will look roughly like
```rust
#[cfg(target_c_char_signed)]
type c_char = i8;
#[cfg(not(target_c_char_signed))]
type c_char = u8;
```

**Comment by Amanieu** [comment]
> Platform `strlen` implementations typically have extensive optimizations, and I don't think we'd want to miss out on those. It might make sense to ensure we have a weak strlen in compiler-builtins or similar. (cc @Amanieu)

I am generally in favor of depending on more libc symbols in core (`strlen`, `memchr`, `bzero`, etc) while providing fallbacks in compiler-builtins for platforms that don't have that symbol.

**Comment by petrochenkov** [comment]
### `trait CStrExt`

Two inherent methods on `CStr` (`into_c_string` and `to_string_lossy`) have signatures including types unavailable in libcore (`CString`, `Cow` and `Box`).

To keep backward compatibility I had to add an unstable `trait CStrExt` with these two stable methods, and add it to prelude.
That trait is only implemented for `CStr`.

### `strlen`

`CStr(ing)` needs to calculate lengths of zero-terminated byte arrays.
While this functionality is trivial, its implementation can also be optimized by using CPU architecture specific techniques.
Such optimized implementations usually live in platform's libc, this is where the current libstd takes them from.

libcore doesn't have access to platform's libc, so it has to either
- Use the naive trivial implementation of `strlen`, or
- Provide the optimized version of `strlen`, similarly to how it does with `memchr`, or
- Use some scheme with weak symbols as suggested above, with weak naive implementation being provided by libcore by default, but then overridable by any other linked libraries, including platform's libc.

**Comment by joshtriplett** [comment]
We discussed this in today's @rust-lang/libs-api team meeting. Many people are enthusiastic to see this work happen.

Regarding relying on target ABI, there were zero objections.

Regarding strlen, there was a consensus in favor of the general approach @Amanieu suggested, of going ahead and using platform C library functions in core as long as they're functions that we can fall back to compiler-builtins for, so that we can use the optimized implementations from the C library if available.

Regarding CStr functions that return types only available in alloc or std: we had some discussion about whether we should use an extension trait for this, or use a lang item or some similar mechanism to allow extending `CStr` from `alloc` and `std`. Our primary concern there was that an extension trait may *technically* be a very slightly breaking change, if someone had a different trait with the same method names; extension trait vs inherent method has different precedence. Using lang items to extend the set of inherent methods from alloc wouldn't have that issue.

Ultimately, though, we decided that we'd rather *not* add lang items for this, and we'd rather use an extension trait as long as that won't break anything in practice. So, summary: let's do a crater run to confirm that the extension trait doesn't break anything, and then go with the extension trait.

**Comment by petrochenkov** [comment]
>we had some discussion about whether we should use an extension trait for this, or use a lang item or some similar mechanism to allow extending `CStr` from `alloc` and `std

When implementing this I recalled that we had such extension traits before, but I couldn't find any of them in libcore now.
Apparently all of them were converted to lang items in 2018 (the main PR was https://github.com/rust-lang/rust/pull/49896), so the existing policy seems to be preferring lang items for this.

This `CStr(ing)` move would require a single new lang item in liballoc.

**Comment by joshtriplett** [comment]
@petrochenkov A lang item seems fine as well. But in that case, I think we need a crater run to confirm no breakage.

**Comment by bstrie** [comment]
Note that https://github.com/rust-lang/rust/pull/94503 has just moved `c_char` (among others) to `core::ffi`.

**Comment by joshtriplett** [comment]
We discussed this again in today's @rust-lang/libs meeting. We're fine with using a lang item in this case, since that's what we already do in other cases.

In general, we'd like to steer away in the future from having any inherent methods on types in core that return types from alloc or std (or inherent methods on types in alloc that return types from std, if any). Such methods make the binding between those two types more special, and make it a little harder to write drop-in replacements for types like CString/String/etc. So, for instance, in this case we should ideally steer people towards `into()` rather than `into_c_string()`, so that people can just `impl From<CStr> for MyCStringType`.

But none of that is a blocker for this move; this method can go ahead and use a lang item as other methods do, and we'll have the more general discussion in the future.

File: library/core/src/ffi.rs [hunk 1]
```diff
@@ -57,6 +62,80 @@ impl fmt::Debug for c_void {
     }
 }
 
+macro_rules! cfg_if {
```
Comment (joshtriplett) [inline library/core/src/ffi.rs:65]: There's now a version of this available (privately) in core.

**Comment by petrochenkov** [comment]
>This `CStr(ing)` move would require a single new lang item in liballoc.

I'm afraid we'll have to keep the extension trait at least for some time for technical reasons.

`cfg(bootstrap)` code (including third-party libraries) uses `CStr::(to_string_lossy,into_c_string)`, so they should be available without any compiler changes until the bootstrap compiler updates.

**Comment by petrochenkov** [comment]
>`cfg(bootstrap)` code (including third-party libraries) uses `CStr::(to_string_lossy,into_c_string)`, so they should be available without any compiler changes until the bootstrap compiler updates.

So I ended up keeping the extension trait only in `cfg(bootstrap)` mode.

**Comment by petrochenkov** [comment]
Blocked on https://github.com/rust-lang/compiler-builtins/pull/457.

**Comment by joshtriplett** [comment]
> > `cfg(bootstrap)` code (including third-party libraries) uses `CStr::(to_string_lossy,into_c_string)`, so they should be available without any compiler changes until the bootstrap compiler updates.
> 
> So I ended up keeping the extension trait only in `cfg(bootstrap)` mode.

Seems like a reasonable transition plan.

**Comment by yaahc** [comment]
looks like there's a test failure

@rustbot author

**Comment by petrochenkov** [comment]
>looks like there's a test failure

Updated with a fix for the `cfg(no_global_oom_handling)` mode.
@rustbot ready

File: compiler/rustc_metadata/src/rmeta/decoder.rs [hunk 2]
```diff
@@ -1752,11 +1753,6 @@ impl CrateMetadata {
             .decode((&blob, sess))
             .map(|trait_impls| (trait_impls.trait_id, trait_impls.impls))
             .collect();
-        let incoherent_impls = root
```
[the remark this answers is not in the document]
Comment (petrochenkov) [reply compiler/rustc_metadata/src/rmeta/decoder.rs:1755]: This is a bugfix for https://github.com/rust-lang/rust/pull/94963 to account for incoherent impl types having `DefId`s, without it this PR ICEs.

**Comment by petrochenkov** [comment]
This needs a fix for targets that don't support `target_has_atomic = "ptr"`.

**Comment by petrochenkov** [comment]
Some wasm-specific dead code warnings.
@rustbot author

**Comment by petrochenkov** [comment]
@lygstate 
This PR implements the strategy from https://github.com/rust-lang/rust/pull/94079#issuecomment-1044688919.
If libc is linked then `strlen` is used from there (this is typically a fast platform-provided implementation), otherwise it's used from compiler-builtins (this is a naive implementation).

If you have issues with this setup, could you open a new issue and describe your case?

**Comment by pnkfelix** [comment]
Visiting for weekly performance triage

* The report above claims this caused a 0.8% regression on primary benchmark unicode-normalization-0.1.19 (check full).
* I was not able to reproduce the regression locally; all my runs showed an improvement.
* This may be running afoul of rust-lang/rustc-perf#1299; so I'm going to chalk this up to noise/undeterminism, rather than a proper regression.

@rustbot label: +perf-regression-triaged

**Comment by CAD97** [comment]
I want to note that this is a regression in rustdoc output, as https://doc.rust-lang.org/nightly/std/ffi/type.CString.html is now a redirect to https://doc.rust-lang.org/nightly/alloc/ffi/struct.CString.html.

Also, should this get a tracking issue rather than pointing to this PR?

**Comment by joshtriplett** [comment]
@CAD97 When we stabilize this, we can switch back from type aliases to re-exports, which should address that.

**Pull request merged**